beacon-chain/rpc/beaconv1/state.go

```diff
@@ -2,29 +2,166 @@ package beaconv1
+	} else {
+		gt, err = ptypes.TimestampProto(genesisTime)
+	}
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not convert genesis time to proto: %v", err)
+	}
+
+	genValRoot := bs.ChainInfoFetcher.GenesisValidatorRoot()
+	return &ethpb.GenesisResponse{
+		GenesisTime:           gt,
+		GenesisValidatorsRoot: genValRoot[:],
+		GenesisForkVersion:    params.BeaconConfig().GenesisForkVersion,
+	}, nil
 }
 
 // GetStateRoot calculates HashTreeRoot for state with given 'stateId'. If stateId is root, same value will be returned.
 func (bs *Server) GetStateRoot(ctx context.Context, req *ethpb.StateRequest) (*ethpb.StateRootResponse, error) {
-	return nil, errors.New("unimplemented")
+	requestedState, err := bs.getState(ctx, req.StateId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get state from ID: %v", err)
+	}
+	stateRoot, err := requestedState.HashTreeRoot(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get state root: %v", err)
+	}
+	return &ethpb.StateRootResponse{
+		StateRoot: stateRoot[:],
+	}, nil
 }
 
 // GetStateFork returns Fork object for state with given 'stateId'.
 func (bs *Server) GetStateFork(ctx context.Context, req *ethpb.StateRequest) (*ethpb.StateForkResponse, error) {
-	return nil, errors.New("unimplemented")
+	requestedState, err := bs.getState(ctx, req.StateId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get state from ID: %f", err)
+	}
+	fork := requestedState.Fork()
+	return &ethpb.StateForkResponse{
+		Fork: &ethpb.Fork{
+			PreviousVersion: fork.PreviousVersion,
+			CurrentVersion:  fork.CurrentVersion,
+			Epoch:           fork.Epoch,
+		},
+	}, nil
 }
 
 // GetFinalityCheckpoints returns finality checkpoints for state with given 'stateId'. In case finality is
 // not yet achieved, checkpoint should return epoch 0 and ZERO_HASH as root.
 func (bs *Server) GetFinalityCheckpoints(ctx context.Context, req *ethpb.StateRequest) (*ethpb.StateFinalityCheckpointResponse, error) {
-	return nil, errors.New("unimplemented")
+	requestedState, err := bs.getState(ctx, req.StateId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get state from ID: %v", err)
+	}
+	prevJustChkpt := requestedState.PreviousJustifiedCheckpoint()
+	curJustChkpt := requestedState.CurrentJustifiedCheckpoint()
+	finalizedChkpt := requestedState.FinalizedCheckpoint()
+	resp := &ethpb.StateFinalityCheckpointResponse{
+		PreviousJustified: &ethpb.Checkpoint{
+			Epoch: prevJustChkpt.Epoch,
+			Root:  prevJustChkpt.Root,
+		},
+		CurrentJustified: &ethpb.Checkpoint{
+			Epoch: curJustChkpt.Epoch,
+			Root:  curJustChkpt.Root,
+		},
+		Finalized: &ethpb.Checkpoint{
+			Epoch: finalizedChkpt.Epoch,
+			Root:  finalizedChkpt.Root,
+		},
+	}
+	return resp, nil
+}
+
+func (bs *Server) getState(ctx context.Context, stateId []byte) (*state.BeaconState, error) {
+	var requestedState *state.BeaconState
```

Comment:
Since this method can be expensive, please set a reasonable timeout. Perhaps 10 seconds?

```suggestion
	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()
	var requestedState *state.BeaconState
```